### ISSUE TITLE
vim-go as an optional package (loaded using packadd)

### ISSUE THREAD

**Issue by aphecetche** [fatih/vim-go#2144]
### What did you do? (required. The issue will be **closed** when not provided.)

For a clean directory (w/o a pre-existing main.go) : 

vim main.go

### What did you expect to happen?

Was expecting to be able to use all `:GoXXX` functions defined by the vim-go plugin.

### What happened instead?

Only seeing `:GoInstallBinaries` `:GoPath` `:GoUpdateBinaries`

Note that if I then do `:set ft=go` then I see all the (other) Go functions.

### Configuration (**MUST** fill this out):

* vim-go version:
master at commit c6537a3fd3dfe26e32ea154d74b7a16c2a63b79f

* `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
filetype plugin on
filetype plugin indent on

* Vim version (first three lines from `:version`):
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Aug 17 2018 17:24:51)
Included patches: 1-503, 505-680, 682-1283
Compiled by root@apple.com

* Go version (`go version`):
go version go1.11.5 darwin/amd64

* Go environment (`go env`):
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/tester/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/tester/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/Cellar/go/1.11.5/libexec"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.11.5/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/pr/7khr_lmj611grlmc7c3b16v80000gq/T/go-build791878719=/tmp/go-build -gno-record-gcc-switches -fno-common"

# Extra info

Note that what I'm trying to achieve is to use `vim-go` as an optional package. Basically in my complete vim setup I have a after/ftplugin directory with one file per "computer language" I'm using. Those files are triggered when a given filetype is used (normal vim behaviour). In those files I load packages and define mappings, functions, etc... that are only relevant for that filetype. 

Relevant here is probably the `.vim/pack/laurent/start/laurent/ftplugin/go.vim` which is (as expected) triggered when opening a .go file (even a new one) and which contains a single line : 

packadd vim-go

(the full struct of my .vim dir for this reproducer is below : ) 

```
.vim
└── pack
    ├── laurent
    │   ├── opt
    │   │   └── laurent
    │   │       ├── plugin
    │   │       │   └── myfunction.vim
    │   │       └── syntax
    │   │           └── laurent.vim
    │   └── start
    │       └── laurent
    │           ├── ftdetect
    │           │   └── laurent.vim
    │           └── ftplugin
    │               ├── go.vim
    │               └── laurent.vim
    └── minpac
        ├── opt
            └── vim-go     
```

Thanks for any help,

Regards,

**Comment by bhcleek** [comment]
vim-go's functions are only loaded when the the filetype is set to `go`. If you're delaying loading filetype plugins until _after_ the filetype is set, then you'll need to set it again to trigger all the behaviors that the filetype plugins provide (e.g. compilers, syntax, commands, mappings, etc.)

**Comment by aphecetche** [comment]
ok, so I'm seeing is a feature, not a bug ;-) 
Must confess I'm still quite confused by all the files involved in vim (yes, even after reading the :help). So I'm probably missing something obvious here , 
but what is the correct way of loading vim-go functionalities only when I'm actually using go files then ?

**Comment by bhcleek** [comment]
It's not _just_ vim-go; it's really any kind of filetype plugin.

> Must confess I'm still quite confused by all the files involved in vim (yes, even after reading the :help).

Understandable; vim can be quite overwhelming at first. I'd definitely suggest starting out slow and try to not be clever 😉. Use the help extensively. You might start by reading `:help init`, `:help rtp`, `:help plugins` and `:help filetype-plugins`.

> but what is the correct way of loading vim-go functionalities only when I'm actually using go files then ?

Just put vim-go in your runtime path or load it using a package manager as described in https://github.com/fatih/vim-go#install. This will cause `:GoInstallBinaries`, `:GoPath`, and `:GoUpdateBinaries` to be available regardless of the filetype, but is that really a problem for you?

I think you might be struggling with understanding when a file is loaded into Vim vs when the functionality provided by that file is available.  You might find `:help :filetype` illuminating.

**Comment by bhcleek** [comment]
@aphecetche are you satisfied, or do you believe this issue should remain open?

**Comment by Carpetsmoker** [comment]
Typing `:e` after `:packadd vim-go` should fix it.

**Comment by bhcleek** [comment]
IMHO, That would make the commands et al. available, but would be clunky and wouldn't be the experience he's really going for.

**Comment by aphecetche** [comment]
Sorry for the late reply. 
My original intent (partly misguided by a fuzzy understanding of how/when vim is loading things) was to "load only what I use". But as a) it does not seem to hurt the launching time and b) makes my life simpler, I've dropped that idea. 
So in the end, I've decided to simply forego the optional loading of packages (i.e. I now put them under some ~/.config/nvim/xxx/start directory instead of ~/.config/nvim/xxx/opt one). 
Thanks for your help,
